### ISSUE TITLE
[ide-diagnostics] unused-variables false-positive

### ISSUE THREAD

**Issue by jhgg** [rust-lang/rust-analyzer#20254]
In the provided code-snippet, `store` is diagnosed as an unused variable by rust-analyzer, but not by rustc.

**rust-analyzer version**: 0.4.2543-standalone (f73ce3c8a8 2025-07-17)

**rustc version**: rustc 1.89.0-nightly (99e7c15e8 2025-06-01)

**editor or extension**: VSCode

**repository link (if public, optional)**: (eg. [rust-analyzer](https://github.com/rust-lang/rust-analyzer))

**code snippet to reproduce**:
```rust
use std::any::Any;

fn main() {
    let store = make();
    //  ^^^^^ unused-variable
    bar(|| foo(store));
    // bar(|| drop(store)) // does not reproduce here.
    // foo(store) // does not reproduce here, either.
}

fn foo(_: Box<dyn Any>) {}
fn bar(_: impl FnOnce()) {}
fn make() -> Box<dyn Any> {
    loop {}
}
```

**Comment by ChayimFriedman2** [comment]
This is somehow because of the definition of `Box`. Making a custom `Box` doesn't replicate the warning. But I don't know what exactly causes it.
